### PR TITLE
Replaces `Free` boilerplate with @freestyle!

### PR DESCRIPTION
**Pull request by raulraja**
**WIP. DO NOT MERGE!**

This is what replacing the current `Free` boilerplate would look like in the Server.
@dialelo @jdesiloniz @juanpedromoreno @rafaparadela @noel @peterneyens thoughts and feedback welcome!

Notice Inject implicits, manual Coproduct alignment, default implicit instances and composed interpreters are no longer required and automatically provided by freestyle. Also notice how interpreters are simple method impls which actual form a FunctionK when extending `WhateverOps.Interpreter[M]`.

More info in this WIP docs as well of what Freestyle does:

https://github.com/47deg/freestyle/blob/master/docs/src/main/tut/algebras.md
https://github.com/47deg/freestyle/blob/master/docs/src/main/tut/modules.md

### CODE REVIEW

**Comment by noelmarkham** [comment]
This looks absolutely awesome, very exciting stuff. No other comments, this is just a great example of the boilerplate reduction given by the library.

**Comment by juanpedromoreno** [comment]
I love this!!

**Comment by jdesiloniz** [comment]
This looks amazing, removing all the boilerplate to use the algebras, hurray!

**Comment by dialelo** [comment]
Awesome stuff, :fire: that inject boilerplate.

File: core/src/main/scala/org/scalaexercises/algebra/progress.scala [hunk 1]
```diff
@@ -10,55 +10,43 @@ import org.scalaexercises.types.user._
...
-) extends UserProgressOp[Option[String]]
+import freestyle._
+import freestyle.implicits._
 
-final case class GetExerciseEvaluations(
-  user:    User,
-  library: String,
-  section: String
-) extends UserProgressOp[List[UserProgress]]
+object imp {
```
Comment (raulraja) [inline core/src/main/scala/org/scalaexercises/algebra/progress.scala]: Not sure why this is necessary. The `Free` companion in cats already includes a monad instance that should satisfy the Applicative requirement

File: server/app/org/scalaexercises/exercises/controllers/ApplicationController.scala [hunk 2]
```diff
@@ -92,7 +96,11 @@ class ApplicationController(cache: CacheApi)(
       library ← exerciseOps.getLibrary(libraryName)
       user ← userOps.getUserByLogin(request.session.get("user").getOrElse(""))
       section ← user.fold(
-        Free.pure[ExercisesApp, Option[String]](None)
+        Free.liftF[FreeApplicative[ExercisesApp.Op, ?], Option[String]](
```
Comment (raulraja) [inline server/app/org/scalaexercises/exercises/controllers/ApplicationController.scala]: `(None: Option[String]).freeS` may work here. 
https://github.com/47deg/freestyle/blob/master/freestyle/shared/src/main/scala/freestyle/package.scala#L99
https://github.com/47deg/freestyle/blob/master/freestyle/shared/src/main/scala/freestyle/package.scala#L29

File: server/app/org/scalaexercises/exercises/services/interpreters/interpreters.scala [hunk 3]
```diff
@@ -153,18 +170,18 @@ trait TestInterpreters extends Interpreters[Id] with IdInstances {
...
 
-  implicit class FreeOps[F[_], A](f: Free[F, A]) {
-    implicit val tailRecMTask = new cats.RecursiveTailRecM[scalaz.concurrent.Task] {}
-
-    def runFuture(implicit interpreter: F ~> Task, T: Transactor[Task], M: Monad[Task]): Future[Either[Throwable, A]] = {
+  implicit class FreeOps[F[_], A](f: FreeS[F, A]) {
+    def runFuture(
+      implicit
+      T: Transactor[Task],
+      M: MonadError[Task, Throwable]
```
Comment (peterneyens) [inline server/app/org/scalaexercises/exercises/services/interpreters/interpreters.scala]: What happens if you add an implicit `F ~> Task` interpreter here?
With `freestyle.implicits._` in the files where you call `.runFuture`.

**Comment by raulraja** [comment]
@dialelo @peterneyens I fixed all remaining Freestyle related compilation issues and right now it's at a point where we just need to bring `Github4s` to latest `cats`

```
[error] missing or invalid dependency detected while loading class file 'Github.class'.
[error] Could not access type RecursiveTailRecM in package cats,
[error] because it (or its dependencies) are missing. Check your build definition for
[error] missing or conflicting dependencies. (Re-run with `-Ylog-classpath` to see the problematic classpath.)
[error] A full rebuild may help if 'Github.class' was compiled against an incompatible version of cats.
```

**Comment by dialelo** [comment]
@raulraja amazing, I'm waiting for [this](https://github.com/47deg/github4s/pull/54) to go in and I'll update this PR when we release github4s with updated dependencies.

**Comment by juanpedromoreno** [comment]
Duplicated by #622

**Pull request closed without merging**